client/topology.go

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/m3db/m3db/encoding"
+	"github.com/m3db/m3db/topology"
+
+	"github.com/m3db/m3x/pool"
+
+	"github.com/uber-go/tally"
+)
+
+func (s *session) initHostQueues(topologyMap topology.Map) ([]hostQueue, int, int, error) {
```

Comment:
Splitting up methods that sit directly on the `session` struct is likely not great, it'll just spread the methods around some files.  

The better approach to refactoring this beast will likely be making it composition based, i.e. have a few new types that handle entire responsibility of some workflows and have the `session` delegate to them.